src/pretrain_utils.py

```diff
@@ -54,37 +63,101 @@ def input_fn_builder(tokenizer: tf_text.BertTokenizer,
+    bin_mask = rearrange(bin_mask, 'channels -> () () channels')
+    target_label = bin_mask * discretized_target
+    target_label = reduce(
+        target_label,
+        'batch seq_len channels -> batch seq_len',
+        'sum')
+    return target_label
+
+  def make_masked_language_model_and_masked_patch_prediction_features(
+    features: Mapping[str, tf.Tensor]) -> Mapping[str, tf.Tensor]:
+    """Randomly mask out text and patch tokens
     Args:
       features: A dictionary of Tensor features, crucially including
           `image_input_ids`, `text_input_ids`.
-
     Returns:
       A new `features` dictionary with global-local transformer side inputs.
     """
 
     features = dict(features)
 
+    # Image
     # image_input_ids: <tf.RaggedTensor>[batch, (patches), (1)].
-    image_input_ids = features['image_input_ids']
-    image_features = tf_text.mask_language_model(
-        image_input_ids, image_item_selector, mask_values_chooser, axis=1)
-
+    img_features = tf_text.mask_language_model(
+        features['image_input_ids'],
+        image_item_selector,
+        mask_values_chooser,
+        axis=1)
+    img_masked_input_ids, img_masked_positions, img_masked_ids = img_features
+
+    # Offset -2 position indices: [CLS] and [PATCH] tokens.
+    _img_masked_positions = (img_masked_positions - 2).to_tensor()
+    batch_size, masked_seq_len = get_shape_list(_img_masked_positions)
+    _img_masked_positions = tf.cast(_img_masked_positions, tf.int32)
+    hs = get_shape_list(features['image_data'])[2]
+    masked_patch_tensor = gather_indexes(features['image_data'],
+                                         _img_masked_positions)
+    masked_patch_tensor = tf.reshape(masked_patch_tensor,
+                                     [batch_size, masked_seq_len, hs])
+    
+    # Create label for masked patches
+    features['masked_patch_target'] = make_masked_patch_prediction_target(
+        masked_patch_tensor)
+
+    # Text
     # text_input_ids: <tf.RaggedTensor>[batch, (words), (wordpieces)].
     text_input_ids = features['text_input_ids']
     if not input_config.mlm_use_whole_word:
       text_input_ids = text_input_ids.merge_dims(-2, -1)
-    text_features = tf_text.mask_language_model(
+    txt_features = tf_text.mask_language_model(
         text_input_ids, text_item_selector, mask_values_chooser, axis=1)
-
+    txt_masked_input_ids, txt_masked_positions, txt_masked_ids = txt_features
+    features['masked_text_target'] = txt_masked_ids
+      
     # Offset text positions after image ones ([CLS] [PATCH] P1 p2 ... P196).
-    text_positions = text_features[1] + 2 + num_patch_per_row ** 2
-    masked_input_ids = tf.concat([image_features[0], text_features[0]], axis=0)
-    masked_positions = tf.concat([image_features[1], text_positions], axis=0)
-    masked_ids = tf.concat([image_features[2], text_features[2]], axis=0)
+    txt_masked_positions = txt_masked_positions + 2 + num_patch_per_row ** 2
+
+    # Join
```

Comment:
Nit: it is better to make the comment more meaningful:) For example, "Join image and text input features...."